File: controller/pessoaCtrl.js
import Endereco from "../model/endereco.js";
import Pessoa from "../model/pessoa.js";

export default class PessoaCtrl {

    gravar(requisicao, resposta) {
        resposta.type("application/json");

        if (requisicao.method === 'POST' && requisicao.is("application/json")) {
            const cpf = requisicao.body.cpf;
            const rg = requisicao.body.rg;
            const nome = requisicao.body.nome;
            const dataNascimento = requisicao.body.dataNascimento;
            const sexo = requisicao.body.sexo;
            const locNascimento = requisicao.body.locNascimento;
            const estadoNascimento = requisicao.body.estadoNascimento;
            const estadoCivil = requisicao.body.estadoCivil;
            const endereco = new Endereco(requisicao.body.endereco.rua, requisicao.body.endereco.numero, requisicao.body.endereco.complemento, requisicao.body.endereco.cep, requisicao.body.endereco.uf, requisicao.body.endereco.cidade);

            if (cpf && rg && nome && dataNascimento && sexo && locNascimento && estadoNascimento && estadoCivil && endereco) {
                const pessoa = new Pessoa(cpf, rg, nome, dataNascimento, sexo, locNascimento, estadoNascimento, estadoCivil, endereco);
                pessoa.consultar(cpf)
                    .then(pessoaExistente => {
                        if (!pessoaExistente) {
                            pessoa.gravar()
                                .then(() => {
                                    resposta.status(200).json({ "status": true, "mensagem": "Pessoa cadastrada com sucesso!" });
                                })
                                .catch(erro => {
                                    resposta.status(500).json({ "status": false, "mensagem": "Erro ao cadastrar a pessoa: " + erro.message });
                                });
                        } else {
                            resposta.status(400).json({ "status": false, "mensagem": "Já existe uma pessoa cadastrada com este CPF." });
                        }
                    })
                    .catch(erro => {
                        resposta.status(500).json({ "status": false, "mensagem": "Erro ao consultar a pessoa: " + erro.message });
                    });
            } else {
                resposta.status(400).json({ "status": false, "mensagem": "Preencha todos os campos corretamente conforme documentação da API." });
            }
        } else {
            resposta.status(400).json({ "status": false, "mensagem": "Requisição inválida! Consulte a documentação da API." });
        }
    }

    editar(requisicao, resposta) {
        resposta.type("application/json");

        if ((requisicao.method === 'PUT' || requisicao.method === 'PATCH') && requisicao.is("application/json")) {
            const cpf = requisicao.body.cpf;
            const rg = requisicao.body.rg;
            const nome = requisicao.body.nome;
            const dataNascimento = requisicao.body.dataNascimento;
            const sexo = requisicao.body.sexo;
            const locNascimento = requisicao.body.locNascimento;
            const estadoNascimento = requisicao.body.estadoNascimento;
            const estadoCivil = requisicao.body.estadoCivil;
            const endereco = new Endereco(requisicao.body.endereco.rua, requisicao.body.endereco.numero, requisicao.body.endereco.complemento, requisicao.body.endereco.cep, requisicao.body.endereco.uf, requisicao.body.endereco.cidade);

            if (cpf && rg && nome && dataNascimento && sexo && locNascimento && estadoNascimento && estadoCivil && endereco) {
                const pessoa = new Pessoa(cpf, rg, nome, dataNascimento, sexo, locNascimento, estadoNascimento, estadoCivil, endereco);
                pessoa.alterar()
                    .then(() => {
                        resposta.status(200).json({ "status": true, "mensagem": "Pessoa atualizada com sucesso!" });
                    })
                    .catch(erro => {
                        resposta.status(500).json({ "status": false, "mensagem": "Erro ao atualizar a pessoa: " + erro.message });
                    });
            } else {
                resposta.status(400).json({ "status": false, "mensagem": "Preencha todos os campos corretamente conforme documentação da API." });
            }
        } else {
            resposta.status(400).json({ "status": false, "mensagem": "Requisição inválida! Consulte a documentação da API." });
        }
    }

    excluir(requisicao, resposta) {
        resposta.type("application/json");

        if (requisicao.method === 'DELETE') {
            const cpf = requisicao.params.cpf;

            if (cpf) {
                const pessoa = new Pessoa(cpf);
                pessoa.apagar()
                    .then(() => {
                        resposta.status(200).json({ "status": true, "mensagem": "Pessoa excluída com sucesso!" });
                    })
                    .catch(erro => {
                        resposta.status(500).json({ "status": false, "mensagem": "Erro ao excluir a pessoa: " + erro.message });
                    });
            } else {
                resposta.status(400).json({ "status": false, "mensagem": "Informe um CPF válido para excluir a pessoa." });
            }
        } else {
            resposta.status(400).json({ "status": false, "mensagem": "Requisição inválida! Consulte a documentação da API." });
        }
    }

    consultar(requisicao, resposta) {
        resposta.type("application/json");

        if (requisicao.method === "GET") {
            const termo = requisicao.params.termo;
            const pessoa = new Pessoa();
            pessoa.consultar(termo)
                .then(listaPessoas => {
                    resposta.status(200).json(listaPessoas);
                })
                .catch(erro => {
                    resposta.status(500).json({ "status": false, "mensagem": "Erro ao consultar a pessoa(s): " + erro.message });
                });
        } else {
            resposta.status(400).json({ "status": false, "mensagem": "Requisição inválida! Consulte a documentação da API." });
        }
    }
}

File: src/main/java/sistema/saga/sagabackend/controller/PessoaCtrl.java
package sistema.saga.sagabackend.controller;

import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import sistema.saga.sagabackend.model.Endereco;
import sistema.saga.sagabackend.model.Pessoa;
import sistema.saga.sagabackend.repository.GerenciaConexao;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class PessoaCtrl {
    public ResponseEntity<Object> gravarPessoa(Map<String, Object> dados) {
        Map<String, Object> resposta = new HashMap<>();
        String cpf = (String) dados.get("cpf");
        String rg = (String) dados.get("rg");
        String nome = (String) dados.get("nome");
        String dataNascimentoStr = (String) dados.get("dataNascimento");
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        LocalDate dataNascimento = LocalDate.parse(dataNascimentoStr, formatter);
        String sexo = (String) dados.get("sexo");
        String locNascimento = (String) dados.get("locNascimento");
        String estadoNascimento = (String) dados.get("estadoNascimento");
        String estadoCivil = (String) dados.get("estadoCivil");
        Map<String, Object> end = (Map<String, Object>) dados.get("endereco");
        String rua = (String) end.get("rua");
        int numero = Integer.parseInt(end.get("numero").toString());
        String complemento = (String) end.get("complemento");
        String cep = (String) end.get("cep");
        String uf = (String) end.get("uf");
        String cidade = (String) end.get("cidade");
        if (verificaIntegridade(cpf) &&
                verificaIntegridade(rg) &&
                verificaIntegridade(nome) &&
                verificaIntegridade(sexo) &&
                verificaIntegridade(locNascimento) &&
                verificaIntegridade(estadoNascimento) &&
                verificaIntegridade(estadoCivil) &&
                verificaIntegridade(rua) &&
                verificaIntegridade(numero) &&
                verificaIntegridade(cep) &&
                verificaIntegridade(uf) &&
                verificaIntegridade(cidade) &&
                verificaIntegridade(dataNascimento)) {
            GerenciaConexao gerenciaConexao;
            try {
                gerenciaConexao = new GerenciaConexao();
                try {
                    gerenciaConexao.getConexao().iniciarTransacao();
                    //begin transaction
                    Endereco endereco = new Endereco(rua, numero, complemento, cep, cidade, uf);
                    if (!endereco.gravar(gerenciaConexao.getConexao())) {
                        resposta.put("status", false);
                        resposta.put("mensagem", "Erro ao cadastrar endereco");
                        //roolback; end trasaction;
                        gerenciaConexao.getConexao().rollback();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.badRequest().body(resposta);
                    }

                    Pessoa pessoa = new Pessoa(
                            cpf,
                            rg,
                            nome,
                            dataNascimento,
                            sexo,
                            locNascimento,
                            estadoNascimento,
                            endereco,
                            estadoCivil
                    );

                    if (pessoa.gravar(gerenciaConexao.getConexao())) {
                        resposta.put("status", true);
                        resposta.put("mensagem", "Pessoa Inserida com sucesso");
                        //commit; end transaction;
                        gerenciaConexao.getConexao().commit();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.ok(resposta);
                    } else {
                        resposta.put("status", false);
                        resposta.put("mensagem", "Pessoa não foi inserida!");
                        //rollback end transaction;
                        gerenciaConexao.getConexao().rollback();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.badRequest().body(resposta);
                    }
                } catch (Exception e) {
                    resposta.put("status", false);
                    resposta.put("mensagem", "Ocorreu um erro na durante a insercao");
                    gerenciaConexao.getConexao().rollback();
                    gerenciaConexao.getConexao().fimTransacao();
                    gerenciaConexao.Desconectar();
                    return ResponseEntity.badRequest().body(resposta);
                }
            } catch (Exception e) {
                resposta.put("status", false);
                resposta.put("mensagem", "Ocorreu um erro ao iniciar conexao");
                return ResponseEntity.badRequest().body(resposta);
            }
        } else {
            resposta.put("status", false);
            resposta.put("mensagem", "Dados inválidos");
            return ResponseEntity.badRequest().body(resposta);
        }

    }

    public ResponseEntity<Object> alterarPessoa(Map<String, Object> dados) {
        Map<String, Object> resposta = new HashMap<>();
        String cpf = (String) dados.get("cpf");
        String rg = (String) dados.get("rg");
        String nome = (String) dados.get("nome");
        String dataNascimentoStr = (String) dados.get("dataNascimento");
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        LocalDate dataNascimento = LocalDate.parse(dataNascimentoStr, formatter);
        String sexo = (String) dados.get("sexo");
        String locNascimento = (String) dados.get("locNascimento");
        String estadoNascimento = (String) dados.get("estadoNascimento");
        String estadoCivil = (String) dados.get("estadoCivil");
        Map<String, Object> end = (Map<String, Object>) dados.get("endereco");
        String rua = (String) end.get("rua");
        int numero = Integer.parseInt(end.get("numero").toString());
        String complemento = (String) end.get("complemento");
        String cep = (String) end.get("cep");
        String uf = (String) end.get("uf");
        String cidade = (String) end.get("cidade");

        if (verificaIntegridade(cpf) &&
                verificaIntegridade(rg) &&
                verificaIntegridade(nome) &&
                verificaIntegridade(sexo) &&
                verificaIntegridade(locNascimento) &&
                verificaIntegridade(estadoNascimento) &&
                verificaIntegridade(estadoCivil) &&
                verificaIntegridade(rua) &&
                verificaIntegridade(numero) &&
                verificaIntegridade(cep) &&
                verificaIntegridade(uf) &&
                verificaIntegridade(cidade) &&
                verificaIntegridade(dataNascimento)) {
            GerenciaConexao gerenciaConexao;
            try {
                gerenciaConexao = new GerenciaConexao();
                try {
                    gerenciaConexao.getConexao().iniciarTransacao();
                    if(complemento==null)
                        complemento="";
                    Endereco endereco = new Endereco(rua, numero, complemento, cep, cidade, uf);
                    if (!endereco.gravar(gerenciaConexao.getConexao())) {
                        resposta.put("status", false);
                        resposta.put("mensagem", "Erro ao cadastrar endereco");
                        gerenciaConexao.getConexao().rollback();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.badRequest().body(resposta);
                    }
                    Pessoa pessoa = new Pessoa(
                            cpf,
                            rg,
                            nome,
                            dataNascimento,
                            sexo,
                            locNascimento,
                            estadoNascimento,
                            endereco,
                            estadoCivil
                    );
                    if (pessoa.alterar(gerenciaConexao.getConexao())) {
                        resposta.put("status", true);
                        resposta.put("mensagem", "Pessoa alterada com sucesso");
                        gerenciaConexao.getConexao().commit();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.ok(resposta);
                    } else {
                        resposta.put("status", false);
                        resposta.put("mensagem", "Pessoa não foi alterada!");
                        gerenciaConexao.getConexao().rollback();
                        gerenciaConexao.getConexao().fimTransacao();
                        gerenciaConexao.Desconectar();
                        return ResponseEntity.badRequest().body(resposta);
                    }
                } catch (Exception e) {
                    resposta.put("status", false);
                    resposta.put("mensagem", "Ocorreu um erro de conexão");
                    gerenciaConexao.getConexao().rollback();
                    gerenciaConexao.getConexao().fimTransacao();
                    gerenciaConexao.Desconectar();
                    return ResponseEntity.badRequest().body(resposta);
                }
            } catch (Exception e) {
                resposta.put("status", false);
                resposta.put("mensagem", "Ocorreu um erro ao iniciar conexao");
                return ResponseEntity.badRequest().body(resposta);
            }
        } else {
            resposta.put("status", false);
            resposta.put("mensagem", "Dados inválidos");
            return ResponseEntity.badRequest().body(resposta);
        }

    }

    public ResponseEntity<Object> apagarPessoa(String cpf) {
        Map<String, Object> resposta = new HashMap<>();

        if (verificaIntegridade(cpf)) {
            try {
                Pessoa pessoa = new Pessoa(cpf);
                GerenciaConexao gerenciaConexao = new GerenciaConexao();
                if (pessoa.apagar(gerenciaConexao.getConexao())) {
                    resposta.put("status", true);
                    resposta.put("mensagem", "Pessoa excluída com sucesso!");
                    gerenciaConexao.Desconectar();
                    return ResponseEntity.ok(resposta);
                } else {
                    resposta.put("status", false);
                    resposta.put("mensagem", "Exclusão não foi realizada!");
                    gerenciaConexao.Desconectar();
                    return ResponseEntity.badRequest().body(resposta);
                }
            } catch (Exception e) {
                resposta.put("status", false);
                resposta.put("mensagem", "Ocorreu um erro de conexão");
                return ResponseEntity.badRequest().body(resposta);
            }
        } else {
            resposta.put("status", false);
            resposta.put("mensagem", "Dados inválidos para exclusão!");
            return ResponseEntity.badRequest().body(resposta);
        }

    }

    public ResponseEntity<Object> buscarTodos() {
        Map<String, Object> resposta = new HashMap<>();
        GerenciaConexao gerenciaConexao = new GerenciaConexao();
        try {
            Pessoa pessoa = new Pessoa();
            List<Map<String,Object>> enderecos = new ArrayList<>();
            List<Pessoa> pessoaList = pessoa.buscarTodos(gerenciaConexao.getConexao(), enderecos);
            if (pessoaList != null ) {
                for (int i = 0; i < enderecos.size(); i++) {
                    Map<String, Object>  end= enderecos.get(i);
                    Endereco endereco = new Endereco(
                            ((Number) end.get("endereco_id")).longValue(),
                            (String) end.get("endereco_rua"),
                            (int) end.get("endereco_num"),
                            (String) end.get("endereco_complemento"),
                            (String) end.get("endereco_cep"),
                            (String) end.get("endereco_cidade"),
                            (String) end.get("endereco_uf")
                    );
                    pessoaList.get(i).setEndereco(endereco);
                }
                resposta.put("status", true);
                resposta.put("listaDePessoas", pessoaList);
                gerenciaConexao.Desconectar();
                return ResponseEntity.ok(resposta);
            } else {
                resposta.put("status", false);
                resposta.put("mensagem", "Não existem pessoas cadastradas");
                gerenciaConexao.Desconectar();
                return ResponseEntity.badRequest().body(resposta);
            }
        } catch (Exception e) {
            resposta.put("status", false);
            resposta.put("mensagem", "Ocorreu um erro de conexão");
            gerenciaConexao.Desconectar();
            return ResponseEntity.badRequest().body(resposta);
        }
    }

    public ResponseEntity<Object> buscarPessoa(String cpf) {
        Map<String, Object> resposta = new HashMap<>();
        GerenciaConexao gerenciaConexao = new GerenciaConexao();
        try {
            Pessoa pessoa = new Pessoa(cpf);
            Map<String, Object> end= new HashMap<>();
            pessoa = pessoa.buscaPessoa(gerenciaConexao.getConexao(), end);
            if (pessoa != null) {
                Endereco endereco = new Endereco(
                        ((Number) end.get("endereco_id")).longValue(),
                        (String) end.get("endereco_rua"),
                        (int) end.get("endereco_num"),
                        (String) end.get("endereco_complemento"),
                        (String) end.get("endereco_cep"),
                        (String) end.get("endereco_cidade"),
                        (String) end.get("endereco_uf")
                );
                pessoa.setEndereco(endereco);
                resposta.put("status", true);
                resposta.put("pessoa", pessoa);
                gerenciaConexao.Desconectar();
                return ResponseEntity.ok(resposta);
            } else {
                resposta.put("status", false);
                resposta.put("mensagem", "Não existem pessoas cadastradas");
                gerenciaConexao.Desconectar();
                return ResponseEntity.badRequest().body(resposta);
            }
        } catch (Exception e) {
            resposta.put("status", false);
            resposta.put("mensagem", "Ocorreu um erro de conexão");
            gerenciaConexao.Desconectar();
            return ResponseEntity.badRequest().body(resposta);
        }
    }

    public ResponseEntity<Object> buscarTodosSemAlunos() {
        Map<String, Object> resposta = new HashMap<>();
        GerenciaConexao gerenciaConexao = new GerenciaConexao();
        try {
            Pessoa pessoa = new Pessoa();
            List<Map<String,Object>> enderecos = new ArrayList<>();
            List<Pessoa> pessoaList = pessoa.buscarTodosSemAlunos(gerenciaConexao.getConexao(), enderecos);
            if (pessoaList != null ) {
                for (int i = 0; i < enderecos.size(); i++) {
                    Map<String, Object>  end= enderecos.get(i);
                    Endereco endereco = new Endereco(
                            ((Number) end.get("endereco_id")).longValue(),
                            (String) end.get("endereco_rua"),
                            (int) end.get("endereco_num"),
                            (String) end.get("endereco_complemento"),
                            (String) end.get("endereco_cep"),
                            (String) end.get("endereco_cidade"),
                            (String) end.get("endereco_uf")
                    );
                    pessoaList.get(i).setEndereco(endereco);
                }
                resposta.put("status", true);
                resposta.put("listaDePessoas", pessoaList);
                gerenciaConexao.Desconectar();
                return ResponseEntity.ok(resposta);
            } else {
                resposta.put("status", false);
                resposta.put("mensagem", "Não existem pessoas cadastradas");
                gerenciaConexao.Desconectar();
                return ResponseEntity.badRequest().body(resposta);
            }
        } catch (Exception e) {
            resposta.put("status", false);
            resposta.put("mensagem", "Ocorreu um erro de conexão");
            gerenciaConexao.Desconectar();
            return ResponseEntity.badRequest().body(resposta);
        }
    }

    private boolean verificaIntegridade(String elemento) {
        return elemento != null && !elemento.trim().isEmpty();
    }

    private boolean verificaIntegridade(int elemento) {
        return elemento > 0;
    }

    private boolean verificaIntegridade(LocalDate elemento) {
        return elemento != null;
    }

    private boolean verificaIntegridade(Endereco elemento) {
        return elemento != null;
    }
}

